### PR TITLE
fix #69 and #70

### LINKED ISSUE #69: deleting a node does not update keys in edge_data
After deleting a vertex, the order of the codes for the other vertices can be reshuffled, such that the "standard" arrangement of vertex labels to identify undirected edges is modified. Yet, the keys of the edge data still follow the previous arrangement of vertex labels. Here is a minimal example.

```julia
julia> using Graphs, MetaGraphsNext

julia> g = MetaGraph(Graph{Int8}(0), Symbol, Int8, Int8, :trial, x -> x, zero(Int8));

julia> add_vertex!(g, :l1, 1); add_vertex!(g, :l2, 2);

julia> add_vertex!(g, :l3, 3); add_vertex!(g, :l4, 4);

julia> code_for(g, :l4) # 4
4

julia> add_edge!(g, :l2, :l1, 10);

julia> g.edge_data
Dict{Tuple{Symbol, Symbol}, Int8} with 1 entry:
  (:l1, :l2) => 10
```

We want to merge `:l4` with `:l1`. First: connect `:l4` to `:l1`'s neighbors, second: delete `:l1`.
```julia
julia> add_edge!(g, :l2, :l4, 10); # connect l4 to l1's only neighbor

julia> g.edge_data
Dict{Tuple{Symbol, Symbol}, Int8} with 2 entries:
  (:l2, :l4) => 10
  (:l1, :l2) => 10

julia> MetaGraphsNext.arrange(g, :l4, :l2)
(:l2, :l4)

julia> g.edge_data[MetaGraphsNext.arrange(g, :l4, :l2)] # edge data is found
10

julia> delete!(g, :l1) # now delete l1 (and its incident edges)

julia> code_for(g, :l4) # 1: has changed. now smaller than the code for l1
1

julia> MetaGraphsNext.arrange(g, :l4, :l2) # different order: because new code for l4
(:l4, :l2)

julia> g.edge_data[MetaGraphsNext.arrange(g, :l4, :l2)] # edge data not found
ERROR: KeyError: key (:l4, :l2) not found
Stacktrace:
 [1] getindex(h::Dict{Tuple{Symbol, Symbol}, Int8}, key::Tuple{Symbol, Symbol})
   @ Base ./dict.jl:484
 [2] top-level scope
   @ REPL[166]:1

julia> for (lab1, lab2) in keys(g.edge_data)
           l1, l2 = MetaGraphsNext.arrange(g, lab1, lab2)
           if lab1 != l1 || lab2 != l2
               @error "labels in edge_data keys = $lab1, $lab2 yet arranged labels = $l1, $l2"
           end
       end
┌ Error: labels in edge_data keys = l2, l4 yet arranged labels = l4, l2
└ @ Main REPL[167]:4
```

### PR DESCRIPTION
**Pull request by cecileane**
#69:
- arrange edge tuples by their vertex labels, to keep `edge_data` accessible after vertex deletion, which changes vertex codes.
- tests added
- documentation: added the need for labels to be comparable with `<`.

#70 for `add_edge!`:
- do nothing if the underlying graph already has an edge
- docstring updated to be explicit that this function cannot be used to update the data of an existing edge.

### CODE REVIEW

**Comment by gdalle** [comment]
Thank you @cecileane! I love people who open issues and immediately after that open PRs to fix them. My favorite kind of users, basically no work for me 🏖️

File: src/directedness.jl [hunk 1]
```diff
@@ -24,7 +26,7 @@ end
 @traitfn function arrange(
     ::MG, label_1, label_2, code_1, code_2
```
Comment (gdalle) [inline src/directedness.jl:27]: We can further simplify by removing these `code_i` argument everywhere in the package

**Review by gdalle — CHANGES_REQUESTED** [review]

**Comment by gdalle** [comment]
Could you maybe add a test specifically checking that the edge data remains the same after vertex deletion? Something similar to #69? Here you are only testing that `arrange` works, which is necessary but less robust

**Comment by gdalle** [comment]
Also, this is a breaking change, could you bump the major version number?

**Comment by cecileane** [comment]
> Could you maybe add a test specifically checking that the edge data remains the same after vertex deletion? Something similar to #69? Here you are only testing that `arrange` works, which is necessary but less robust

I added tests in that direction, but I'm not sure I understood what you mean. The `edge_data` is modified in place (entries removed) so the main concern was about the keys, not the values (edge data).

File: src/directedness.jl (hunk 1 above)
Comment (cecileane) [reply src/directedness.jl]: I was thinking that too, but that meant small changes in many places. Done now!

**Comment by bramtayl** [comment]
I'm not sure I like the API of doing nothing if the vertex already exists. I think it makes more sense to change the data if the user requests it?

**Comment by cecileane** [comment]
> I'm not sure I like the API of doing nothing if the vertex already exists.

Do you mean if the _edge_ already exists, and we try to add it again with `add_edge!`, possibly with different data than the currently existing data for that edge?

I don't have any strong opinion. It might be "safe" to not overwrite existing data with `add_edge!` simply because "add" suggests that the edge is absent before being added. In that case, modifying an edge data should be done with `graph[label1, label2] = data`. But this API choice is your choice, not mine. I'm happy to update the PR either way.

**Comment by gdalle** [comment]
> The edge_data is modified in place (entries removed) so the main concern was about the keys, not the values (edge data).

I just meant that we should check we are still able to access the right data for each edge after removal.

> In that case, modifying an edge data should be done with graph[label1, label2] = data

Maybe `add_edge!` should just error when the edge already exists

**Comment by bramtayl** [comment]
> Do you mean if the edge already exists, and we try to add it again with add_edge!, possibly with different data than the currently existing data for that edge?

Yup

> I don't have any strong opinion

Me neither

**Comment by cecileane** [comment]
> Maybe `add_edge!` should just error when the edge already exists

I think `add_edge!(::MetaGraph)` should follow Graphs' choice where `Graphs.add_edge!(::Graph)` returns false if the edge addition failed, with no error.

**Comment by cecileane** [comment]
done with b98373a: to make `add_edge!` modify existing edge data.

File: src/directedness.jl (hunk 1 above)
Comment (cecileane) [reply src/directedness.jl]: by the way: this simplification was done.
I also added a couple tests to check that the edge data was accessible and as expected.

File: src/graphs.jl [hunk 2]
```diff
@@ -174,18 +174,18 @@ Add an edge `(label_1, label_2)` to MetaGraph `meta_graph` with metadata `data`.
 If the `EdgeData` type of `meta_graph` is `Nothing`, `data` can be omitted.
 
 Return `true` if the edge has been added, `false` otherwise.
-If `(label_1, label_2)` already existed, its data is *not* updated to `data`.
+If `(label_1, label_2)` already existed, its data is updated to `data`.
```
Comment (gdalle) [inline src/graphs.jl:177]: and false is returned

Comment (gdalle) [reply src/graphs.jl:177]:
```suggestion
If `(label_1, label_2)` already existed, its data is updated to `data` and `false` is returned nonetheless.
```

File: test/misc.jl [hunk 3]
```diff
@@ -8,6 +8,11 @@ function test_labels_codes(mg::MetaGraph)
     for (label_1, label_2) in edge_labels(mg)
         @test has_edge(mg, code_for(mg, label_1), code_for(mg, label_2))
     end
+    # below: arrange(edges) ⊆ keys of mg.edge_data. then = because same length
+    for e in edge_labels(mg)
+        @test_logs mg[e...]  # no log, no error
```
Comment (gdalle) [inline test/misc.jl:13]: What does this mean?

Comment (cecileane) [reply test/misc.jl:13]: It means that this line only tests for the absence of any log record (e.g. no warning via `@warn`) and no Exception / error.

An alternative test could check the output of `arrange` is a key of `edge_data`, but the current test checks the API that users would use rather than the lower-level data structures.

Comment (gdalle) [reply test/misc.jl:13]: sounds good to me

**Review by gdalle — APPROVED** [review]
